hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/snapshot/TestSnapshotDeletingService.java

```diff
@@ -482,70 +683,143 @@ private void createSnapshotDataForBucket1() throws Exception {
+        bucket.getName() + "key1", false);
     // Key 2 is deleted here, which will be reclaimed here as
     // it is not being referenced by previous snapshot.
-    client.getProxy().deleteKey(VOLUME_NAME, BUCKET_NAME_ONE,
-        "bucket1key2", false);
-    client.getProxy().deleteKey(VOLUME_NAME, BUCKET_NAME_ONE,
-        "bucket1key0", false);
+    client.getProxy().deleteKey(bucket.getVolumeName(), bucket.getName(),
+        bucket.getName() + "key2", false);
+    client.getProxy().deleteKey(bucket.getVolumeName(), bucket.getName(),
+        bucket.getName() + "key0", false);
     assertTableRowCount(keyTable, 0);
     // one copy of bucket1key0 should also be reclaimed as it not same
     // but original deleted key created during overwrite should not be deleted
     assertTableRowCount(deletedTable, 2);
 
     // Create Snapshot 2.
-    client.getProxy().createSnapshot(VOLUME_NAME, BUCKET_NAME_ONE,
-        "bucket1snap2");
+    client.getProxy().createSnapshot(bucket.getVolumeName(), bucket.getName(),
+        bucket.getName() + "snap2");
     assertTableRowCount(snapshotInfoTable, 2);
     // Key 2 is removed from the active Db's
     // deletedTable when Snapshot 2 is taken.
     assertTableRowCount(deletedTable, 0);
 
-    TestDataUtil.createKey(bucket1, "bucket1key3", ReplicationFactor.THREE,
+    TestDataUtil.createKey(bucket, bucket.getName() + "key3", ReplicationFactor.THREE,
         ReplicationType.RATIS, CONTENT);
-    TestDataUtil.createKey(bucket1, "bucket1key4", ReplicationFactor.THREE,
+    TestDataUtil.createKey(bucket, bucket.getName() + "key4", ReplicationFactor.THREE,
         ReplicationType.RATIS, CONTENT);
-    client.getProxy().deleteKey(VOLUME_NAME, BUCKET_NAME_ONE,
-        "bucket1key4", false);
+    client.getProxy().deleteKey(bucket.getVolumeName(), bucket.getName(),
+        bucket.getName() + "key4", false);
     assertTableRowCount(keyTable, 1);
     assertTableRowCount(deletedTable, 0);
 
     // Create Snapshot 3.
-    client.getProxy().createSnapshot(VOLUME_NAME, BUCKET_NAME_ONE,
-        "bucket1snap3");
+    client.getProxy().createSnapshot(bucket.getVolumeName(), bucket.getName(),
+        bucket.getName() + "snap3");
     assertTableRowCount(snapshotInfoTable, 3);
 
     SnapshotInfo snapshotInfo = metadataManager.getSnapshotInfoTable()
-        .get("/vol1/bucket1/bucket1snap2");
+        .get(String.format("/%s/%s/%ssnap2", bucket.getVolumeName(), bucket.getName(), bucket.getName()));
 
     // Delete Snapshot 2.
-    client.getProxy().deleteSnapshot(VOLUME_NAME, BUCKET_NAME_ONE,
-        "bucket1snap2");
+    client.getProxy().deleteSnapshot(bucket.getVolumeName(), bucket.getName(),
+        bucket.getName() + "snap2");
     assertTableRowCount(snapshotInfoTable, 2);
-    verifySnapshotChain(snapshotInfo, "/vol1/bucket1/bucket1snap3");
+    verifySnapshotChain(snapshotInfo, String.format("/%s/%s/%ssnap3", bucket.getVolumeName(), bucket.getName(),
+        bucket.getName()));
   }
 
+
```

Comment:
Can we have a test case where we have a type snapshot chain such that some keys are reclaimable, some keys are getting renamed multiple times between snapshots, and some of those keys are also common with delete table? Then we let the snapshot deletion service run on this chain and make sure that AOS has the correct merge of all these keys. These asserts at various deletion stages would make a one comprehensive test case.

Reply:
This is already covered as part of this unit test case. Integration test is not required here. https://github.com/apache/ozone/blob/7124b1eb1bb9d372ba7506318bbea38b67e75599/hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/response/snapshot/TestOMSnapshotMoveTableKeysResponse.java#L109